### PR TITLE
Fix BlockHound false positives

### PR DESCRIPTION
**Pull request by dkhalanskyjb**

### CODE REVIEW

File: kotlinx-coroutines-debug/src/CoroutinesBlockHoundIntegration.kt [hunk 1]
```diff
@@ -1,16 +1,53 @@
...
 
 @Suppress("UNUSED")
-public class CoroutinesBlockHoundIntegration: BlockHoundIntegration {
+public class CoroutinesBlockHoundIntegration : BlockHoundIntegration {
 
-    override fun applyTo(builder: BlockHound.Builder) {
-        builder.addDynamicThreadPredicate { isSchedulerWorker(it) }
-        builder.nonBlockingThreadPredicate { p -> p.or { mayNotBlock(it) } }
+    override fun applyTo(builder: BlockHound.Builder): Unit = with(builder) {
+        // These classes use a lock internally, but should be safe to use.
```
Comment (qwwdfsad) [inline kotlinx-coroutines-debug/src/CoroutinesBlockHoundIntegration.kt:13]: I would rather write something a bit more encouraging for external users such as "These classes use a lock internally, but only for well-known, short and isolated state transitions that are known to be short and fast"

File: kotlinx-coroutines-debug/src/CoroutinesBlockHoundIntegration.kt [hunk 2]
```diff
@@ -1,16 +1,53 @@
...
+        for (method in listOf(
```
Comment (qwwdfsad) [inline kotlinx-coroutines-debug/src/CoroutinesBlockHoundIntegration.kt:14]: Currently, it's a bit hard to read and get through. Could you please split it to a few methods?
Like
```
allowBlockingCallsInsideChannelImpl()
allowBlockingCallsInsideReflection()
... etc. ...
```

**Review by qwwdfsad — CHANGES_REQUESTED** [review]
Please revisit a few more places on our code that rely on fast locking:
- `StateFlow` and `SharedFlow`
- `JobSupport` (`finalizeFinishingState` & co.)

**Comment by dkhalanskyjb** [comment]
Yeah, I completely missed that some code is protected using `synchronized`.

File: kotlinx-coroutines-debug/src/CoroutinesBlockHoundIntegration.kt [hunk 3]
```diff
@@ -1,16 +1,145 @@
...
+    /** Allows blocking inside [kotlinx.coroutines.channels.ConflatedChannel]. */
+    private fun BlockHound.Builder.allowBlockingCallsInConflatedChannel() {
+        for (method in listOf("offerInternal", "offerSelectInternal", "pollInternal", "pollSelectInternal",
+            "onCancelIdempotent"))
+        {
+            allowBlockingCallsInside("kotlinx.coroutines.channels.ConflatedChannel", method)
+        }
+    }
+
+    /** Allows blocking when enqueuing tasks into a thread pool.
```
Comment (qwwdfsad) [inline kotlinx-coroutines-debug/src/CoroutinesBlockHoundIntegration.kt:106]: In our codestyle, we typically start comment on the next line after `/**`

**Review by qwwdfsad — APPROVED** [review]
Otherwise looks good, great job

**Pull request merged**